shared/src/main/scala/kinesis4cats/producer/Producer.scala

```diff
@@ -172,39 +176,40 @@ abstract class Producer[F[_], PutReq, PutRes] private[kinesis4cats] (
     *   a [[cats.data.NonEmptyList NonEmptyList]] of
     *   [[kinesis4cats.producer.Record Records]]
     * @return
-    *   [[cats.data.Ior Ior]] with the following:
-    *   - left: a [[kinesis4cats.producer.Producer.Error Producer.Error]], which
-    *     represents records that were too large to be put on kinesis as well as
-    *     records that were not produced successfully in the batch request.
-    *   - right: a [[cats.data.NonEmptyList NonEmptyList]] of the underlying put
-    *     responses
+    *   Producer.Result
     */
-  def put(records: NonEmptyList[Record]): F[Producer.Res[PutRes]] = {
+  def put(records: NonEmptyList[Record]): F[Producer.Result[PutRes]] = {
     val ctx = LogContext()
 
     for {
       ref <- Ref.of(Producer.RetryState[PutRes](records, None))
       finalRes <- retryingOnFailuresAndAllErrors(
         config.retryPolicy,
-        (x: Producer.Res[PutRes]) =>
-          F.pure(x.isRight || x.left.exists(e => e.errors.exists(_.isLeft))),
-        (x: Producer.Res[PutRes], details: RetryDetails) =>
+        (x: Producer.Result[PutRes]) =>
+          F.pure(x.isSuccessful || (x.isPartiallySuccessful && !x.hasFailed)),
+        (x: Producer.Result[PutRes], details: RetryDetails) =>
           for {
-            failedRecords <- F.fromOption(
-              x.left.flatMap { e =>
-                e.errors.flatMap(errors => errors.right)
-              },
+            failed <- F.fromOption(
+              NonEmptyList.fromList(x.failed),
               new RuntimeException(
                 "Failed records empty, this should never happen"
               )
             )
             _ <- logger.debug(ctx.addEncoded("retryDetails", details).context)(
-              s"Failures with ${failedRecords.size} records detected, retrying failed records"
+              s"Failures with ${failed.length} records detected, retrying failed records"
             )
             _ <- ref.update(current =>
               Producer.RetryState(
-                failedRecords.map(_.record),
-                current.res.fold(Some(x))(y => Some(y.combine(x)))
+                failed.map(_.record),
+                current.res.fold(Some(x)) { currentResult =>
+                  Some(
+                    Producer.Result(
+                      currentResult.successful ++ x.successful,
+                      currentResult.invalid ++ x.invalid,
+                      failed.toList // Only use failed from most recent result
```

Comment:
Allows for the final result to be indicative of the remaining failed records rather than all records that failed during the retry loop.